src/installed.rs

```
@@ -0,0 +1,357 @@
+    {
+        let result: Result<String, Box<Error>> = match self.server {
+            None => {
+                let url = build_authentication_request_url(&appsecret.auth_uri,
+                                                           &appsecret.client_id,
+                                                           scopes,
+                                                           None);
+                match auth_delegate.present_user_url(&url, true /* need_code */) {
+                    None => {
+                        Result::Err(Box::new(io::Error::new(io::ErrorKind::UnexpectedEof,
+                                                            "couldn't read code")))
+                    }
+                    // Remove newline
+                    Some(mut code) => {
+                        code.pop();
+                        Result::Ok(code)
+                    }
+                }
+            }
+            Some(_) => {
+                // The redirect URI must be this very localhost URL, otherwise Google refuses
+                // authorization.
+                let url = build_authentication_request_url(&appsecret.auth_uri,
+                                                           &appsecret.client_id,
+                                                           scopes,
+                                                           Some(format!("http://localhost:{}",
+                                                                        self.port
+                                                                            .unwrap_or(8080))));
+                auth_delegate.present_user_url(&url, false /* need_code */);
+
+                match self.auth_code_rcv.as_ref().unwrap().recv() {
+                    Result::Err(e) => Result::Err(Box::new(e)),
+                    Result::Ok(s) => Result::Ok(s),
+                }
+            }
+        };
+        let _ = self.server.as_mut().map(|l| l.close());
```

well, `ok()` doesn't work because it's already returning an option. But `is_some()` suppresses the error too (I guess that's what your intention is)